src/sabre_layout.rs

```diff
@@ -0,0 +1,207 @@
+    neighbor_table: &NeighborTable,
+    distance_matrix: PyReadonlyArray2<f64>,
+    heuristic: &Heuristic,
+    seed: Option<u64>,
+    max_iterations: usize,
+    num_swap_trials: usize,
+    num_layout_trials: usize,
+) -> ([NLayout; 2], SwapMap, PyObject) {
+    let run_in_parallel = getenv_use_multiple_threads();
+    let outer_rng = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+    let seed_vec: Vec<u64> = outer_rng
+        .sample_iter(&rand::distributions::Standard)
+        .take(num_layout_trials)
+        .collect();
+    let dist = distance_matrix.as_array();
+    let result = if run_in_parallel && num_layout_trials > 1 {
+        seed_vec
+            .into_par_iter()
+            .enumerate()
+            .map(|(index, seed_trial)| {
+                (
+                    index,
+                    layout_trial(
+                        num_clbits,
+                        &mut dag_nodes.clone(),
+                        neighbor_table,
+                        &dist,
+                        heuristic,
+                        seed_trial,
+                        max_iterations,
+                        num_swap_trials,
+                        run_in_parallel,
+                    ),
+                )
+            })
+            .min_by_key(|(index, result)| {
+                (
+                    result.1.map.values().map(|x| x.len()).sum::<usize>(),
+                    *index,
+                )
+            })
+            .unwrap()
+            .1
+    } else {
+        seed_vec
+            .into_iter()
+            .map(|seed_trial| {
+                layout_trial(
+                    num_clbits,
+                    &mut dag_nodes,
+                    neighbor_table,
+                    &dist,
+                    heuristic,
+                    seed_trial,
+                    max_iterations,
+                    num_swap_trials,
+                    run_in_parallel,
+                )
+            })
+            .min_by_key(|result| result.1.map.values().map(|x| x.len()).sum::<usize>())
+            .unwrap()
+    };
+    (result.0, result.1, result.2.into_pyarray(py).into())
+}
+
+fn layout_trial(
+    num_clbits: usize,
+    dag_nodes: &mut Vec<(usize, Vec<usize>, HashSet<usize>)>,
+    neighbor_table: &NeighborTable,
+    distance_matrix: &ArrayView2<f64>,
+    heuristic: &Heuristic,
+    seed: u64,
+    max_iterations: usize,
+    num_swap_trials: usize,
+    run_swap_in_parallel: bool,
+) -> ([NLayout; 2], SwapMap, Vec<usize>) {
+    // Pick a random initial layout and fully populate ancillas in that layout too
+    let num_physical_qubits = distance_matrix.shape()[0];
+    let mut rng = Pcg64Mcg::seed_from_u64(seed);
+    let mut physical_qubits: Vec<usize> = (0..num_physical_qubits).collect();
+    physical_qubits.shuffle(&mut rng);
+    let mut phys_to_logic = vec![0; num_physical_qubits];
+    physical_qubits
```

Comment:
Not particularly, I think I just did it this way because of the enumerate, but I'll convert it to a for loop (along with your suggestion below to have a custom constructor for this form).

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/9116/commits/bf9bef8a8959fecceb05cc8862a42ca829af2c92